### ISSUE TITLE
problem to activate RBAC

### ISSUE THREAD

**Issue by ericmagna** [strimzi/strimzi-kafka-oauth#99]
Hi,

I downloaded the last version to configure Kafka in oauth2 authentication with the keycloak provider. This configuration works fine.

Now i try to configure KeycloakRBACAuthorizer without succes because i got this error 
ERROR [KafkaServer id=1721704] Fatal error during KafkaServer startup. Prepare to shutdown (kafka.server.KafkaServer)
java.lang.RuntimeException: Specified token endpoint uri is invalid: "https://<fdqn keycloak>/auth/realms/sandbox_kafka/protocol/openid-connect/token" 

I try with this lines 
######################### Enabling the KeycloakRBACAuthorizer #########################
authorizer.class.name=io.strimzi.kafka.oauth.server.authorizer.KeycloakRBACAuthorizer
principal.builder.class=io.strimzi.kafka.oauth.server.OAuthKafkaPrincipalBuilder

strimzi.authorization.token.endpoint.uri="https://<fqdn keycloak>/auth/realms/sandbox_kafka/protocol/openid-connect/token"
strimzi.authorization.client.id="<client id>"
strimzi.authorization.ssl.truststore.location="kafka.server.truststore.jks"
strimzi.authorization.ssl.truststore.password="<password>"
strimzi.authorization.ssl.truststore.type="jks"
strimzi.authorization.ssl.endpoint.identification.algorithm=""
strimzi.authorization.kafka.cluster.name=cluster2
strimzi.authorization.delegate.to.kafka.acl=false

I have attached the complete configuration files (server.properites and kafka_jaas_conf).

I can't figure out why it is putting this java exception with this configuration.

Thank you for your help.

[strimzi_rbac.zip](https://github.com/strimzi/strimzi-kafka-oauth/files/6130500/strimzi_rbac.zip)

**Comment by scholzj** [comment]
@mstruk Any idea?

**Comment by mstruk** [comment]
Well, the error pretty much spells it out :)

The url for the token endpoint is not a valid url - it has an empty hostname.
Additionally, you have to set `strimzi.authorization.client.id` to a meaningful value - it cannot be empty. You have to set it to the client ID of the OAuth2 client configured to represent your Kafka broker - in our docs and examples we mostly use `kafka` for this.
Also, I suggest you use an absolute path for `strimzi.authorization.ssl.truststore.location`.

**Comment by ericmagna** [comment]
i verify that url is correct by a curl command with certifcate and it's work. 

I use the same url without disabled  rbac config to use authentication OAuth2 with provider keycloak. No problem.

With RBAC enabled, I got this error 
java.lang.RuntimeException: Specified token endpoint uri is invalid: "https://<fqdn_keycloak>/auth/realms/sandbox_kafka/protocol/openid-connect/token"
with this config
authorizer.class.name=io.strimzi.kafka.oauth.server.authorizer.KeycloakRBACAuthorizer
principal.builder.class=io.strimzi.kafka.oauth.server.OAuthKafkaPrincipalBuilder

strimzi.authorization.token.endpoint.uri="https://<fqdn keycloak>/auth/realms/sandbox_kafka/protocol/openid-connect/token"
strimzi.authorization.client.id="kafka-broker"
strimzi.authorization.ssl.truststore.location="/jks/kafka.server.truststore.jks"
strimzi.authorization.ssl.truststore.password="<password>"
strimzi.authorization.ssl.truststore.type="jks"
strimzi.authorization.ssl.endpoint.identification.algorithm=""
strimzi.authorization.kafka.cluster.name=cluster2
strimzi.authorization.delegate.to.kafka.acl=false

i put an client id but not client secret. I don't find any configuration for client secret in RBAC configuration. it's normal ?

**Comment by mstruk** [comment]
Can you provide a full stacktrace?

**Comment by ericmagna** [comment]
Extract error from log :
[2021-03-13 00:12:21,793] ERROR [KafkaServer id=1721704] Fatal error during KafkaServer startup. Prepare to shutdown (kafka.server.KafkaServer)
java.lang.RuntimeException: Specified token endpoint uri is invalid: "https://keycloak.mutual69.k8s.recouv/auth/realms/sandbox_kafka/protocol/openid-connect/token"
        at io.strimzi.kafka.oauth.server.authorizer.KeycloakRBACAuthorizer.configure(KeycloakRBACAuthorizer.java:190)
        at kafka.server.KafkaServer.$anonfun$startup$5(KafkaServer.scala:288)
        at scala.Option.map(Option.scala:163)
        at kafka.server.KafkaServer.startup(KafkaServer.scala:286)
        at kafka.server.KafkaServerStartable.startup(KafkaServerStartable.scala:38)
        at kafka.Kafka$.main(Kafka.scala:75)
        at kafka.Kafka.main(Kafka.scala)

Attach Files with conf + log complete
[strimzirbac.zip](https://github.com/strimzi/strimzi-kafka-oauth/files/6133624/strimzirbac.zip)

**Comment by mstruk** [comment]
Your error happens [here in the code](https://github.com/strimzi/strimzi-kafka-oauth/blob/0.7.1/oauth-keycloak-authorizer/src/main/java/io/strimzi/kafka/oauth/server/authorizer/KeycloakRBACAuthorizer.java#L190).

Which means that the url you have specified in the configuration is invalid due to some invalid character or an invalid structure.

In fact, it looks like quotes are passed as part of the url. You should not use quotes for `strimzi.authorizer.*` properties.

Also your `server.properties` is invalid in other ways.

You specify `strimzi.authorizer.delegate.class.name=io.strimzi.kafka.oauth.server.authorizer.KeycloakRBACAuthorizer` which has nothing to do with `KeycloakRBACAuthorizer` and is ignored.

You are missing:
```
listener.name.client.oauthbearer.sasl.jaas.config=org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule required ;
```

Please, have a look at [the docs](https://github.com/strimzi/strimzi-kafka-oauth/blob/0.7.1/README.md)

**Comment by ericmagna** [comment]
I removed the double quotes and the rbac works with an authorization denial.

This refusal seems to be linked to the super user that I am going to analyze.

jaas is configured in the kafka_jaas_conf file and I check your remark in my jaas file.

for the line _strimzi.authorizer.delegate.class.name = io.strimzi.kafka.oauth.server.authorizer.KeycloakRBACAuthorizer_, you are right the documentation does not refer to RBAC.

however in the compose-auth.yml file I found this line KAFKA_STRIMZI_AUTHORIZER_DELEGATE_CLASS_NAME: io.strimzi.kafka.oauth.server.authorizer.KeycloakRBACAuthorizer. Since I had errors, I mistakenly thought that this line was missing. 

Thank you for your clarifications and your help.

I can create a PR to add precision for double quotes and fixed the docker example for compose-authz.yml if you want.
